src/scheduler/scheduler.rs

```diff
@@ -261,121 +236,188 @@ impl<VM: VMBinding> GCWorkScheduler<VM> {
     }
 
     #[cold]
     fn poll_slow(&self, worker: &GCWorker<VM>) -> Box<dyn GCWork<VM>> {
-        debug_assert!(!worker.shared.is_parked());
-        let mut guard = self.worker_monitor.0.lock().unwrap();
         loop {
-            debug_assert!(!worker.shared.is_parked());
-            if let Some((work, bucket_is_empty)) = self.pop_scheduable_work(worker) {
-                if bucket_is_empty {
-                    worker
-                        .sender
-                        .send(CoordinatorMessage::BucketDrained)
-                        .unwrap();
-                }
+            let guard = self.worker_monitor.0.lock().unwrap();
+            // Retry polling
+            if let Some(work) = self.poll_schedulable_work(worker) {
                 return work;
             }
-            // Park this worker
-            worker.shared.parked.store(true, Ordering::SeqCst);
-            if self.all_workers_parked() {
-                worker
-                    .sender
-                    .send(CoordinatorMessage::AllWorkerParked)
-                    .unwrap();
+            // Prepare to park this worker
+            let all_parked = self.worker_group.inc_parked_workers();
+            // If all workers are parked, try activate new buckets
+            if all_parked {
+                // If there're any designated work, resume the workers and process them
+                if self.worker_group.has_designated_work() {
+                    self.worker_group.dec_parked_workers();
+                    self.worker_monitor.1.notify_all();
+                    continue;
+                }
+                if self.update_buckets() {
+                    // We're not going to sleep since a new bucket is just open.
+                    self.worker_group.dec_parked_workers();
+                    // We guarantee that we can at least fetch one packet.
+                    let work = self.poll_schedulable_work(worker).unwrap();
+                    // Optimize for the case that a newly opened bucket only has one packet.
+                    // We only notify_all if there're more than one packets available.
+                    if !self.all_activated_buckets_are_empty() {
+                        // Have more jobs in this buckets. Notify other workers.
+                        self.worker_monitor.1.notify_all();
+                    }
+                    // Return this packet and execute it.
+                    return work;
+                }
+                debug_assert!(!self.worker_group.has_designated_work());
+                // The current pause is finished if we can't open more buckets.
+                self.pending_messages.fetch_add(1, Ordering::SeqCst);
+                worker.sender.send(CoordinatorMessage::Finish).unwrap();
             }
             // Wait
-            guard = self.worker_monitor.1.wait(guard).unwrap();
+            let _guard = self.worker_monitor.1.wait(guard).unwrap();
```

Comment:
This `wait` may wake up spuriously.  It may not be a problem because the condition to wake up is "there are pending work packets to do", and it is notified when new work packets are added.  However, there are memory order issues.  See my other comments regarding `WorkBucket::is_activated`.